File: clients/python/clausters/gui/notation/engraver.py
"""The engraver: a score in, a ``score`` display list out.

This is the client-side rendering step. `Score` holds a loaded document so it
can be **edited** and re-engraved; `engrave` is the one-shot for a page that
will not change; `svg_to_display_list` is the adapter both flow through, split
out so it is testable on a captured SVG and shared with every other client.
`page_json` is how a re-engraved page replaces the one on screen.

The engraver is libverovio bound in Rust (``clausters-notation``) and the walk
itself is `clausters_core::notation`, both reached through the C ABI — so this
module is a shell of names, dicts and one handle whose lifetime Python owns,
and a client in another language rebinds the same ABI instead of
reimplementing any of it.
"""

from __future__ import annotations

import json
import os

from ... import _libpath, _native
from ._abi import _MISSING, _engraver, _text, _u8
from .mei import from_notes, from_timeline

# Where the SMuFL data the engraver reads lives. verovio bakes a resource path
# in at *build* time -- the configure-time prefix, which is not where a wheel's
# copy ends up -- so an installed package has to say where its own staged data
# is, or every engraving fails with "font resources are not available". The
# native side reads `CLAUSTERS_VEROVIO` and looks for `<dir>/verovio` under it;
# `setdefault` so an explicit override still wins, and only when the bundled
# directory is actually there (a source checkout has none and falls back to the
# build prefix baked in by `clausters-notation`'s build script).
if os.path.isdir(os.path.join(_libpath.LIBS_DIR, "verovio")):
    os.environ.setdefault("CLAUSTERS_VEROVIO", _libpath.LIBS_DIR)

# The display-list keys the host draws from — everything but `notes`, which is
# the client's own layer. `page_json` and `guidef.score` send exactly these.
_PAGE_LAYERS = ("vb", "glyphs", "prims", "cursors", "step")

class Score:
    """A loaded score, kept alive so it can be **edited** and re-engraved.

    `engrave` is the one-shot form — load, draw, discard. This is the stateful
    one: it holds the engraver's document open, so an edit can be applied to the
    same one the display list was drawn from and the page re-engraved against it.
    The MEI ``xml:id``s survive editing, which is what lets the host keep its
    selection across the round trip: the id the user clicked still names the same
    note afterwards.

    The edit cycle and the undo stack are the shared layer's
    (``clausters_notation::Score``), not this shell's: every edit runs the
    editor action, then a commit that re-runs the layout, then a reload that
    refreshes the MIDI/timemap cache an edit does not invalidate; undo is a stack
    of MEI snapshots, because the engraver's own stack cannot survive that reload
    and its ``undo`` on an empty stack takes the process down. What Python owns
    is the **handle's lifetime**: the score is freed when this object is.
    """

    def __init__(self, data: str, *, scale: int = 40, page_width: int = 2100,
                 options: dict | None = None):
        lib = _engraver()
        extra = json.dumps(options).encode("utf-8") if options else b""
        raw = data.encode("utf-8")
        self._h = lib.clausters_score_open(
            _native.as_u8(raw), len(raw), scale, page_width,
            _native.as_u8(extra), len(extra))
        if not self._h:
            raise ValueError("the engraver could not load the score data")

    def __del__(self):
        handle, self._h = getattr(self, "_h", None), None
        if handle:
            _native.lib().clausters_score_free(handle)

    def display_list(self, page: int = 1) -> dict:
        """This score engraved into a ``score`` display list — the same layers
        `engrave` returns, but from the live document, so it reflects every edit
        applied so far."""
        return json.loads(_text(_native.lib().clausters_score_display_list,
                                self._h, page))

    def mei(self) -> str:
        """The score as MEI, ids and all — the format to persist, and what the
        undo stack is made of."""
        return _text(_native.lib().clausters_score_mei, self._h)

    @property
    def can_undo(self) -> bool:
        return bool(_native.lib().clausters_score_can_undo(self._h))

    @property
    def can_redo(self) -> bool:
        return bool(_native.lib().clausters_score_can_redo(self._h))

    def undo(self) -> bool:
        """Step back one edit. False (never a crash) when there is nothing to
        undo."""
        return bool(_native.lib().clausters_score_undo(self._h))

    def redo(self) -> bool:
        """Step forward again after `undo`; False when there is nothing to redo."""
        return bool(_native.lib().clausters_score_redo(self._h))

    def transpose(self, element_id: str, steps: int) -> bool:
        """Move a note by ``steps`` **diatonic** steps along the staff — up when
        positive — as one undo step.

        This is the pitch edit as the *engraver* expresses it, in steps rather
        than in a position, because its coordinate-taking ``drag`` reads an
        absolute page y in a frame that does not line up with the display list's
        (passing a note its own drawn y moves it six steps), so a caller would
        have to carry an unexplained offset. Steps are exact.

        It is **not** the shape an edit travels in — a displacement made against
        a page since re-engraved would have to be rebased. `transpose_to` is
        what applies what a host sends; reach for this one only when the delta
        is what you actually have.
        """
        raw = element_id.encode("utf-8")
        return bool(_native.lib().clausters_score_transpose(
            self._h, _native.as_u8(raw), len(raw), steps))

    def transpose_to(self, element_id: str, position: int, page: int = 1) -> bool:
        """Move a note **to** the diatonic staff position ``position`` on
        ``page`` — whole steps from its staff's top line, positive upward — as
        one undo step.

        The absolute form, and what a ``"transpose"`` edit-back from the GUI
        host carries: applying it twice leaves the note where it is, and a page
        re-engraved under the gesture needs no rebasing. The relative call
        underneath is the engraver's requirement, and the delta is computed
        against the engraving rather than carried from wherever the gesture
        happened — which is the point, since the two can differ.

        Host and engraver read the position off the same drawing, so a position
        named by one and resolved by the other cannot mean two things.

        Returns whether the note is now at ``position`` — **True when it was
        already there**, since the requested state holds and a resend must be
        harmless. False when the element is not on that page, the page has no
        staff to measure against, or the engraver refused the move.
        """
        raw = element_id.encode("utf-8")
        return bool(_native.lib().clausters_score_transpose_to(
            self._h, _native.as_u8(raw), len(raw), position, page))

    def edit(self, action: str, **param) -> bool:
        """Apply one raw editor action (``set``, ``insert``, ``delete``, ...) as
        a single undo step — the escape hatch for what `transpose` does not
        cover. Returns whether the engraver accepted it; a rejected action leaves
        the score untouched."""
        act = action.encode("utf-8")
        par = json.dumps(param).encode("utf-8")
        return bool(_native.lib().clausters_score_edit(
            self._h, _native.as_u8(act), len(act),
            _native.as_u8(par), len(par)))

    @classmethod
    def from_notes(cls, notes, *, meter: str = "4/4", clef: str = "G2",
                   key: str = "C", beat_unit: int = 4, **kw) -> "Score":
        """An editable `Score` built from a **monophonic** run of events — the
        `from_notes` encoder handed straight to the constructor. ``kw`` passes
        ``scale``/``page_width`` through. See `from_notes` for the mapping."""
        return cls(from_notes(notes, meter=meter, clef=clef, key=key,
                              beat_unit=beat_unit), **kw)

    @classmethod
    def from_timeline(cls, timeline, *, meter: str = "4/4", clef: str = "G2",
                      key: str = "C", beat_unit: int = 4, **kw) -> "Score":
        """An editable `Score` built from a `Timeline` (chords from simultaneous
        events, rests from gaps) — the `from_timeline` encoder handed to the
        constructor. ``kw`` passes ``scale``/``page_width`` through."""
        return cls(from_timeline(timeline, meter=meter, clef=clef, key=key,
                                 beat_unit=beat_unit), **kw)


def engrave(data: str, *, page: int = 1, scale: int = 40,
            page_width: int = 2100, options: dict | None = None) -> dict:
    """Engrave ``data`` (a score in any format the engraver auto-detects) into a
    ``score`` display list.

    One-shot: the score is loaded, drawn and discarded. Use `Score` instead when
    the page has to be **edited** and redrawn.

    The result holds one engraving, in three layers:

    - what the host **draws** — ``vb`` (the ``[w, h]`` page-unit viewBox),
      ``glyphs`` (a SMuFL codepoint-to-outline table), ``prims`` (the placed
      glyphs, lines, fills and texts) and ``step`` (page units per diatonic
      step, the quantum a pitch drag on the page counts in);
    - where the **cursor** goes — ``cursors``, the timemap folded into geometry
      (``{"t", "x", "y0", "y1"}`` per onset, ``t`` in ms);
    - what **sounds** — ``notes``, one ``{"t", "dur", "pitch", "id"}`` per note
      (ms and MIDI pitch). This layer stays on the client: it is what a driver
      plays, and playing it while anchoring the widget's ``playhead_at`` to the
      sample clock of that instant puts the cursor on the sounding note. The
      engraver mints fresh ids per load, so all three layers must come from one
      engraving — which is why one call produces them all.

    Pass the result to `clausters.gui.guidef.score` as its ``display_list``, or
    to `score_view` to get a scrollable page; the builder sends only the drawing
    and cursor layers. The score **wraps into systems** at ``page_width`` (page
    units), and the page grows as tall as the music needs (all systems on one
    page), so a long score reads at ``scale`` instead of being squeezed onto one
    line. ``scale`` sets the staff size; extra engraver ``options`` are merged
    over the defaults. Raises ``RuntimeError`` if the engraver is not built in.
    """
    return Score(data, scale=scale, page_width=page_width,
                 options=options).display_list(page)


def page_json(display_list: dict) -> str:
    """The **drawing** layers of ``display_list`` as the JSON string a live
    ``GuiHost.set(score_id, display_list=…)`` takes — how a re-engraved page
    replaces the one on screen after an edit, without redefining the window.

    The same layers `clausters.gui.guidef.score` sends when it builds the
    widget, so the widget looks the same either way: the client-side ``notes``
    stay here, and so does the host's own chrome (the playhead and the
    selection survive the replacement, which is what keeps the edited note
    selected across the round trip).
    """
    return json.dumps({k: display_list[k] for k in _PAGE_LAYERS
                       if k in display_list})


def svg_to_display_list(svg: str) -> dict:
    """Walk an engraver SVG string into a ``score`` display list. Split out of
    `engrave` so it is testable on a captured SVG, and shared with every other
    client — the walk itself is `clausters_core::notation`, reached through the
    ABI, so a wasm client feeding a wasm engraver's SVG gets the identical list.

    Each primitive carries the id of the element it belongs to, and a **sounding
    element owns everything drawn inside it**: the engraver identifies a note's
    stem and flag separately, and collapsing them onto the note's id is what
    makes one note one thing to select and drag. A chord keeps its notes
    distinct, so one of them can still be transposed alone.
    """
    if not _native.has_notation():
        raise RuntimeError(_MISSING)
    raw = svg.encode("utf-8")
    out = _text(_native.lib().clausters_core_svg_to_display_list,
                _native.as_u8(raw), len(raw))
    return json.loads(out) if out else {}

File: clients/web/src/gui/notation/engraver.ts
// The engraver and its output (mirrors `clausters/gui/notation/engraver.py`).
//
// A score is laid out into SVG by verovio and that SVG is walked into the flat,
// resolution-independent display list the host's `score` widget consumes. Both
// steps are the shared core's: the walk is `clausters_core::notation`, and the
// stateful, editable document is `clausters_core::notation::Score` — the order
// an edit is made in, the reload that keeps the timemap honest, the undo stack.
// What is here is the shell: idiomatic names, plain objects, and the handle's
// lifetime.
//
// The engraver is the pinned verovio compiled to wasm (`_verovio.ts`), the same
// sources and the same options the native library is built and configured with,
// so a page and a window engrave one score into one drawing.

import {
    Score as CoreScore,
    engraveOptions,
    svgToDisplayList as coreSvgToDisplayList,
} from "../../core/clausters_core_web.js";
import { Toolkit } from "./_verovio.ts";
import { Editing } from "../editing/context.ts";
import type { Intent } from "../../document.ts";
import { fromNotes, fromTimeline } from "./mei.ts";
import type { MeiOptions } from "./mei.ts";
import type { Op, Sheet } from "./sheet.ts";
import type { Event as SeqEvent } from "../../seq/event.ts";
import type { Timeline } from "../../seq/timeline.ts";

/**
 * The display-list keys the host draws from — everything but `notes`, which is
 * the client's own layer. {@link pageJson} and the `score` builder send exactly
 * these.
 */
const PAGE_LAYERS = ["vb", "glyphs", "prims", "cursors", "step", "elements"] as const;

/** One engraved page: what is drawn, where the cursor goes, and what sounds. */
export interface Page {
    /** The `[w, h]` page-unit viewBox. */
    vb: number[];
    /** A SMuFL codepoint-to-outline table. */
    glyphs: Record<string, unknown>;
    /** The placed glyphs, lines, fills and texts. */
    prims: Record<string, unknown>[];
    /** Page units per diatonic step — the quantum a pitch drag counts in. */
    step: number;
    /** The timemap folded into geometry: `{t, x, y0, y1}` per onset, `t` in ms. */
    cursors: Record<string, number>[];
    /** One `{t, dur, pitch, id}` per note (ms and MIDI pitch). Client-side. */
    notes: { t: number; dur: number; pitch: number; id: string }[];
    [key: string]: unknown;
}

/** How a score is engraved: the staff size, the wrap width, extra options. */
export interface EngraveOptions {
    /** Staff size (verovio `scale`). */
    scale?: number;
    /** The page units the score wraps into systems at. */
    pageWidth?: number;
    /** Extra engraver options, merged over the defaults. */
    options?: Record<string, unknown> | null;
}

/**
 * A loaded score, kept alive so it can be **edited** and re-engraved.
 *
 * {@link engrave} is the one-shot form — load, draw, discard. This is the
 * stateful one: it holds the engraver's document open, so an edit can be applied
 * to the same one the display list was drawn from and the page re-engraved
 * against it. The MEI `xml:id`s survive editing, which is what lets the host keep
 * its selection across the round trip: the id the user clicked still names the
 * same note afterwards.
 *
 * The edit cycle is the shared layer's (`clausters_core::notation::Score`), not
 * this shell's — as it is in the Python client, which holds the same model over
 * the C ABI.
 *
 * **The undo order is the editing context's**, like every other editable
 * structure's. A score registers in `Editing.of(score)` under the `"score"`
 * vocabulary and records each edit as the MEI it produced, with the previous one
 * as its inverse — an absolute payload, so a step is idempotent and carries no
 * direction. That is what makes a window holding a lane and a page walk **one**
 * order: before this, an engraved page had a real history of its own and Ctrl+Z
 * meant one of two different things depending on what the pointer was over.
 *
 * The shared layer's own snapshot stack is still there and is what a caller with
 * no such context uses — a standalone host holding a page and nothing else. From
 * here it is not read: {@link Score.undo} and {@link Score.redo} walk the
 * context's pile and put a state back through the crate's `load`.
 *
 * **Opening is asynchronous where the Python client's is not**, and only because
 * of where verovio is: a page loads the engraver module the first time one is
 * needed, and a page may not block. Once open, every method is synchronous, as
 * there it is.
 */
export class Score {
    private readonly inner: CoreScore;
    private readonly toolkit: Toolkit;

    private constructor(inner: CoreScore, toolkit: Toolkit) {
        this.inner = inner;
        this.toolkit = toolkit;
    }

    /**
     * Load `data` — a score in any format the engraver auto-detects — and keep
     * the document open.
     */
    static async open(
        data: string,
        { scale = 40, pageWidth = 2100, options = null }: EngraveOptions = {},
    ): Promise<Score> {
        const configured = JSON.parse(
            engraveOptions(scale, pageWidth, options ? JSON.stringify(options) : undefined),
        ) as Record<string, unknown>;
        const toolkit = await Toolkit.open(configured);
        try {
            return new Score(new CoreScore(toolkit as unknown as object, data), toolkit);
        } catch (error) {
            toolkit.free();
            throw error;
        }
    }

    /**
     * An editable score built from a **monophonic** run of events — the
     * {@link fromNotes} encoder handed straight to {@link Score.open}.
     */
    static fromNotes(
        notes: Iterable<SeqEvent>,
        options: MeiOptions & EngraveOptions = {},
    ): Promise<Score> {
        return Score.open(fromNotes(notes, options), options);
    }

    /**
     * An editable score built from a `Timeline` (chords from simultaneous
     * events, rests from gaps).
     */
    static fromTimeline(
        timeline: Timeline,
        options: MeiOptions & EngraveOptions = {},
    ): Promise<Score> {
        return Score.open(fromTimeline(timeline, options), options);
    }

    /**
     * This score engraved into a page — from the live document, so it reflects
     * every edit applied so far.
     */
    displayList(page = 1): Page {
        return JSON.parse(this.inner.displayList(page)) as Page;
    }

    /** The score as MEI, ids and all — the format to persist. */
    mei(): string {
        return this.inner.mei();
    }

    /**
     * The open score as the **model**.
     *
     * Every one of the model's verbs applies to it, including on a score that was
     * typed rather than built: the engraver normalizes whatever it loaded and
     * this reads that, so a phrase in ABC is as editable as one made by
     * operating on a motif.
     *
     * Throws when the document could not be read into a model — a state and not
     * a failure, since the page still draws and still plays and only the
     * model's verbs are unavailable on it.
     */
    sheet(): Sheet {
        return JSON.parse(this.inner.sheet()) as Sheet;
    }

    /**
     * Apply one **model** operation as a single undo step, and re-engrave.
     *
     * This is the edit path. `op` is the payload the sheet verbs build, so an
     * edit to an open score and an edit to a sheet in hand are the same
     * operation through the same code — which is what lets a standalone host
     * with no client language perform it too.
     *
     * `false` when the document has no model behind it or the operation was
     * refused; either way the page and the model are as they were.
     */
    apply(op: Op): boolean {
        const before = this.mei();
        if (!this.inner.apply(JSON.stringify(op))) return false;
        this.record(before, String((op as { op?: unknown }).op ?? "edit the score"));
        return true;
    }

    // ---- the editing context: one order over everything being edited ----

    /**
     * This score's editing context, attaching it on first ask.
     *
     * The context hangs on the **data**, so a page opened twice is one structure
     * in one order; attaching makes the score a participant, which is what lets
     * an undo made in a *neighbouring* window step it too.
     */
    private get editing(): Editing {
        const context = Editing.of(this);
        context.attach(this);
        return context;
    }

    /** This score's identity in the pile — minted once, per score. */
    private get structure(): number {
        return this.editing.identity(this, "score");
    }

    /**
     * Record one edit: the MEI it produced, and the one it replaced.
     *
     * **A state, not a step.** The page a score is at describes it whole, the
     * way a curve's points and a timeline's events do, so an entry is idempotent
     * and reads the same in both directions — which is the only shape a pile
     * shared with other structures can carry.
     */
    private record(before: string, label: string): void {
        const after = this.mei();
        if (after === before) return; // a resend is not an edit
        const context = this.editing;
        context.turn(this, () => {
            // `forward` is a **step** and `backward` is a bare payload: the pile
            // hands an inverse back as what it holds, and wrapping it a second
            // time hands the reader a leg it cannot read.
            context.history.record(
                [
                    {
                        structure: this.structure,
                        // The payload is this domain's own vocabulary and
                        // the pile never reads it; the arrangement's `Intent` is
                        // only what the type happens to name.
                        forward: { edit: { mei: after } as unknown as Intent },
                        backward: { mei: before },
                    },
                ],
                { label },
            );
            context.changed();
        });
    }

    /**
     * Put back the legs of a history step that name **this** score.
     *
     * What the context hands round. A page is not a view — nothing here redraws
     * — so a caller re-engraves after a step exactly as it does after an edit.
     */
    projectLegs(legs: readonly unknown[]): boolean {
        let moved = false;
        for (const leg of legs as { structure?: number; payload?: { mei?: unknown } }[]) {
            if (Math.trunc(Number(leg.structure ?? -1)) !== this.structure) continue;
            const mei = leg.payload?.mei;
            if (typeof mei !== "string") continue;
            moved = this.load(mei) || moved;
        }
        return moved;
    }

    /**
     * Another window in this context edited. Nothing here: a score is data and
     * draws nothing of its own — whoever engraved the page redraws it.
     */
    adopt(): void {}

    /**
     * Replace the document with `mei` — **a state, not a step**.
     *
     * The door the pile puts a previous page back through. It clears the shared
     * layer's own stack, so one score has one history.
     */
    load(mei: string): boolean {
        return this.inner.load(mei);
    }

    /**
     * Whether the **context's** pile has an edit to step back over — which may
     * be an edit to something else entirely, since the order is one.
     */
    get canUndo(): boolean {
        return this.editing.history.canUndo;
    }

    /** @see {@link Score.canUndo} */
    get canRedo(): boolean {
        return this.editing.history.canRedo;
    }

    /**
     * Step the editing context back one edit, and say whether anything moved.
     * `false` when there is nothing to undo.
     *
     * The step is the **context's**, not this score's: a page edited beside a
     * lane steps in the order the two were edited in, and a leg naming another
     * structure is projected by whoever holds it.
     */
    undo(): boolean {
        return this.step("undo");
    }

    /** Step forward again after {@link Score.undo}; `false` when there is nothing to redo. */
    redo(): boolean {
        return this.step("redo");
    }

    private step(direction: "undo" | "redo"): boolean {
        const context = this.editing;
        const legs = context.step(direction);
        return legs === undefined ? false : context.distribute(legs, this);
    }

    /**
     * Move the note `elementId` by `steps` **diatonic** steps along the staff —
     * up when positive — as one undo step.
     *
     * It is the **model's** move where the page named a model item (the note
     * takes the key signature's alteration for the letter it lands on, which is
     * what reading in a key means) and falls back to the engraver's editor only
     * for an element this layer did not write. The relative form; reach for it
     * only when the delta is what you actually have.
     */
    transpose(elementId: string, steps: number): boolean {
        const before = this.mei();
        if (!this.inner.transpose(elementId, steps)) return false;
        this.record(before, "transpose");
        return true;
    }

    /**
     * Move the note `elementId` **to** the diatonic staff position `position` on
     * `page` — the shape an edit travels in, so a resend cannot move the note
     * twice. True when the note is now there, including when it already was.
     */
    transposeTo(elementId: string, position: number, page = 1): boolean {
        const before = this.mei();
        if (!this.inner.transposeTo(elementId, position, page)) return false;
        // **A note already there is not an edit.** The call answers true for it,
        // because the requested state holds; `record` sees the same MEI on both
        // sides and leaves the pile alone, which is the rule a resend follows
        // everywhere else.
        this.record(before, "transpose");
        return true;
    }

    /**
     * Apply one raw editor action (`set`, `insert`, `delete`, …) as a single undo
     * step — the escape hatch for what {@link Score.transpose} does not cover.
     * A rejected action leaves the score untouched.
     */
    edit(action: string, param: Record<string, unknown> = {}): boolean {
        const before = this.mei();
        if (!this.inner.edit(action, JSON.stringify(param))) return false;
        this.record(before, action);
        return true;
    }

    /** The engraver's version — what both clients must agree on. */
    engraverVersion(): string {
        return this.toolkit.version();
    }

    /**
     * Free the document and its toolkit. A page that keeps engraving keeps the
     * module: what is released here is one score, not the engraver.
     */
    free(): void {
        this.inner.free();
        this.toolkit.free();
    }
}

/**
 * Engrave `data` (a score in any format the engraver auto-detects) into a
 * display list.
 *
 * One-shot: the score is loaded, drawn and discarded. Use {@link Score} instead
 * when the page has to be **edited** and redrawn.
 *
 * The result holds one engraving in three layers — what the host **draws**
 * (`vb`, `glyphs`, `prims`, `step`), where the **cursor** goes (`cursors`, the
 * timemap folded into geometry) and what **sounds** (`notes`, which stays on the
 * client: it is what a driver plays). The engraver mints fresh ids per load, so
 * all three must come from one engraving — which is why one call produces them
 * all.
 */
export async function engrave(
    data: string,
    { page = 1, ...options }: EngraveOptions & { page?: number } = {},
): Promise<Page> {
    const score = await Score.open(data, options);
    try {
        return score.displayList(page);
    } finally {
        score.free();
    }
}

/**
 * The **drawing** layers of a display list, as the object a live
 * `host.set(scoreId, { displayList })` takes — how a re-engraved page replaces
 * the one on screen after an edit, without redefining the window.
 *
 * The same layers the `score` builder sends when it builds the widget, so the
 * widget looks the same either way: the client-side `notes` stay here, and so
 * does the host's own chrome (the playhead and the selection survive the
 * replacement, which is what keeps the edited note selected across the round
 * trip).
 */
export function pageJson(displayList: Page): Record<string, unknown> {
    const out: Record<string, unknown> = {};
    for (const key of PAGE_LAYERS) {
        if (key in displayList) out[key] = displayList[key];
    }
    return out;
}

/**
 * Walk an engraver SVG into a display list. Split out of {@link engrave} so it
 * is testable on a captured SVG, and shared with every other client — the walk
 * itself is `clausters_core::notation`, so a page feeding it the same SVG a
 * window feeds it gets the identical list.
 *
 * Each primitive carries the id of the element it belongs to, and a **sounding
 * element owns everything drawn inside it**: the engraver identifies a note's
 * stem and flag separately, and collapsing them onto the note's id is what makes
 * one note one thing to select and drag. A chord keeps its notes distinct, so one
 * of them can still be transposed alone.
 */
export function svgToDisplayList(svg: string): Record<string, unknown> {
    const out = coreSvgToDisplayList(svg);
    return out ? (JSON.parse(out) as Record<string, unknown>) : {};
}
